plugins/integration/providers/voicebase/lib/KVoicebaseIntegrationEngine.class.php

```diff
@@ -0,0 +1,99 @@
+<?php
+/**
+ * @package plugins.voicebase
+ * @subpackage Scheduler
+ */
+class KVoicebaseIntegrationEngine implements KIntegrationCloserEngine
+{
+	private $baseEndpointUrl = null;
+	private $clientHelper = null;
+	
+	/* (non-PHPdoc)
+	 * @see KIntegrationCloserEngine::dispatch()
+	 */
+	public function dispatch(KalturaBatchJob $job, KalturaIntegrationJobData &$data)
+	{
+		return $this->doDispatch($job, $data, $data->providerData);
+	}
+	
+	/* (non-PHPdoc)
+	 * @see KIntegrationCloserEngine::close()
+	 */
+	public function close(KalturaBatchJob $job, KalturaIntegrationJobData &$data)
+	{
+		return $this->doClose($job, $data, $data->providerData);
+	}
+	
+	protected function doDispatch(KalturaBatchJob $job, KalturaIntegrationJobData &$data, KalturaVoicebaseJobProviderData $providerData)
+	{
+		$entryId = $providerData->entryId;
+		$flavorAssetId = $providerData->flavorAssetId;
+		$spokenLanguage = $providerData->spokenLanguage;
+		$formatsString = $providerData->captionAssetFormats;
+		$formatsArray = explode(',', $formatsString);
+
+		$shouldReplaceRemoteMedia = $providerData->replaceMediaContent;
+		$fileLocation = $providerData->fileLocation;
+		$ks = $providerData->ksForExternalService; 
+	
+		$callBackUrl = $data->callbackNotificationBaseUrl;
+		$callBackUrl .= "/id/" . $job->id . "/ks/$ks";
+	
+		$serviceProviderParams = array('apiKey' => $providerData->apiKey , 'apiPassword' => $providerData->apiPassword);
+		$this->clientHelper = VoicebasePlugin::getClientHelper($serviceProviderParams);
+	
+		$flavorAssetId = $this->validateFlavorAssetId($entryId, $flavorAssetId);
+		$flavorUrl = KBatchBase::$kClient->flavorAsset->getUrl($flavorAssetId);
+	
+		$externalEntryExists = $this->clientHelper->checkExitingExternalContent($entryId);
+		if (!$externalEntryExists)
+		{
+			$uploadSuccess = $this->clientHelper->uploadMedia($flavorUrl, $entryId, $callBackUrl, $spokenLanguage, $fileLocation);
+			if(!$uploadSuccess)
+				throw new Exception("upload failed");
+		}
+		elseif($shouldReplaceRemoteMedia == KalturaNullableBoolean::TRUE_VALUE)
+		{
+			$this->clientHelper->deleteRemoteFile($entryId);
+			$uploadSuccess = $this->clientHelper->uploadMedia($flavorUrl, $entryId, $callBackUrl, $spokenLanguage, $fileLocation);
+			if(!$uploadSuccess)
+				throw new Exception("upload failed");
+		}
+		elseif($fileLocation)
+		{
+			$this->clientHelper->updateRemoteTranscript($entryId, $fileLocation, $callBackUrl);
+		}	
+		else
+		{
+			return true;
+		}
+
+		return false;
+	}
+	
+	protected function doClose(KalturaBatchJob $job, KalturaIntegrationJobData &$data, KalturaVoicebaseJobProviderData $providerData)
+	{
+		return false;
```

Comment:
if you're referring to the job's timeout , it's being checked by the KAsyncIntegrateCloser before implementing this engine.